full-service/src/bin/main.rs

```diff
@@ -56,7 +56,11 @@ async fn main() -> Result<(), rocket::Error> {
         exit(EXIT_INVALID_HOST);
     }
 
-    let (logger, _global_logger_guard) = create_app_logger(o!());
+    let (logger, global_logger_guard) = create_app_logger(o!());
+
+    // This is necessary to prevent the logger from being reset when it goes out of
+    // scope so that rocket can use it in its own async context
+    global_logger_guard.cancel_reset();
```

Comment:
How does the logger go out of scope if its inside main()?

Reply:
It's only at the end when you shut down the app that this happens, but it definitely does crash

Reply:
This isn't the real main loop, it's actually a special rocket async main function under the hood

Reply:
If you're curious, this is the error I get

```
OH NO, WE CRASHED :( thread main on /home/briancorbin/.mobilecoin/test/wallet-service-mirror-public
Details: panicked at 'slog-scope: No logger set. Use `slog_scope::set_global_logger` or `slog_scope::scope`.', /tmp/mobilenode/.mob/cargo/registry/src/github.com-1ecc6299db9ec823/slog-scope-4.4.0/lib.rs:125:13
   0:     0x558c7d76d6d7 - backtrace::backtrace::libunwind::trace::hf08cd3e760f923ad
                               at /tmp/mobilenode/.mob/cargo/registry/src/github.com-1ecc6299db9ec823/backtrace-0.3.63/src/backtrace/libunwind.rs:93:5
                           backtrace::backtrace::trace_unsynchronized::hd426b87efd9f024c
                               at /tmp/mobilenode/.mob/cargo/registry/src/github.com-1ecc6299db9ec823/backtrace-0.3.63/src/backtrace/mod.rs:66:5
                           backtrace::backtrace::trace::h47b03002f8d2d5e2
                               at /tmp/mobilenode/.mob/cargo/registry/src/github.com-1ecc6299db9ec823/backtrace-0.3.63/src/backtrace/mod.rs:53:14
   1:     0x558c7d759725 - backtrace::capture::Backtrace::create::h093e91b295ca9dd2
                               at /tmp/mobilenode/.mob/cargo/registry/src/github.com-1ecc6299db9ec823/backtrace-0.3.63/src/capture.rs:176:9
                           backtrace::capture::Backtrace::new::hee432b17cb854b1e
                               at /tmp/mobilenode/.mob/cargo/registry/src/github.com-1ecc6299db9ec823/backtrace-0.3.63/src/capture.rs:140:22
   2:     0x558c7d38c474 - mc_common::panic_handler::handle_panic::ha793082fbf1d7fd4
                               at /tmp/mobilenode/mobilecoin/common/src/panic_handler.rs:20:38
                           mc_common::panic_handler::setup_panic_handler::{{closure}}::h80b944243d725038
                               at /tmp/mobilenode/mobilecoin/common/src/panic_handler.rs:14:9
   3:     0x558c7dbdff26 - std::panicking::rust_panic_with_hook::h9ade200275992b5b
                               at /rustc/e85edd9a844b523a02dbd89f3c02cd13e4c9fe46/library/std/src/panicking.rs:702:17
   4:     0x558c7d7b32ab - std::panicking::begin_panic::{{closure}}::h8ab6bdc0776513c1
                               at /rustc/e85edd9a844b523a02dbd89f3c02cd13e4c9fe46/library/std/src/panicking.rs:617:9
   5:     0x558c7d7b3274 - std::sys_common::backtrace::__rust_end_short_backtrace::hf63175cb8b41b6c7
                               at /rustc/e85edd9a844b523a02dbd89f3c02cd13e4c9fe46/library/std/src/sys_common/backtrace.rs:138:18
   6:     0x558c7d09f3ec - std::panicking::begin_panic::h3b6a8f6faf87f190
                               at /rustc/e85edd9a844b523a02dbd89f3c02cd13e4c9fe46/library/std/src/panicking.rs:616:12
   7:     0x558c7d7b2f09 - <slog_scope::NoGlobalLoggerSet as slog::Drain>::log::h4a312839de8b7c52
                               at /tmp/mobilenode/.mob/cargo/registry/src/github.com-1ecc6299db9ec823/slog-scope-4.4.0/lib.rs:125:13
   8:     0x558c7d3c84c7 - <alloc::sync::Arc<D> as slog::Drain>::log::h133922607de653bc
                               at /tmp/mobilenode/.mob/cargo/registry/src/github.com-1ecc6299db9ec823/slog-2.7.0/src/lib.rs:1719:9
                           slog::Logger<D>::log::h0d2c7faecf6d185b
                               at /tmp/mobilenode/.mob/cargo/registry/src/github.com-1ecc6299db9ec823/slog-2.7.0/src/lib.rs:1233:17
                           <slog_stdlog::Logger as log::Log>::log::{{closure}}::h863c34815ad44d56
                               at /tmp/mobilenode/.mob/cargo/registry/src/github.com-1ecc6299db9ec823/slog-stdlog-4.1.1/lib.rs:109:42
                           slog_scope::with_logger::{{closure}}::hfc03c35259a84a90
                               at /tmp/mobilenode/.mob/cargo/registry/src/github.com-1ecc6299db9ec823/slog-scope-4.4.0/lib.rs:221:21
                           std::thread::local::LocalKey<T>::try_with::hcad035aaec3d0e2f
                               at /rustc/e85edd9a844b523a02dbd89f3c02cd13e4c9fe46/library/std/src/thread/local.rs:443:16
                           std::thread::local::LocalKey<T>::with::h004657b51f79ca67
                               at /rustc/e85edd9a844b523a02dbd89f3c02cd13e4c9fe46/library/std/src/thread/local.rs:419:9
   9:     0x558c7d3c80d7 - slog_scope::with_logger::h2d2b2401fd4201fa
                               at /tmp/mobilenode/.mob/cargo/registry/src/github.com-1ecc6299db9ec823/slog-scope-4.4.0/lib.rs:217:5
                           <slog_stdlog::Logger as log::Log>::log::h926e8a55cac2c0ce
                               at /tmp/mobilenode/.mob/cargo/registry/src/github.com-1ecc6299db9ec823/slog-stdlog-4.1.1/lib.rs:109:9
  10:     0x558c7dbbe536 - log::__private_api_log::h17593d5d02028faa
                               at /tmp/mobilenode/.mob/cargo/registry/src/github.com-1ecc6299db9ec823/log-0.4.11/src/lib.rs:1374:5
  11:     0x558c7d298f8b - rocket::config::config::Config::pretty_print::h1a7e55ab8ff449e7
                               at /tmp/mobilenode/.mob/cargo/registry/src/github.com-1ecc6299db9ec823/rocket-0.5.0-rc.2/src/config/config.rs:357:9
  12:     0x558c7d0f5a16 - rocket::rocket::Rocket<rocket::phase::Build>::ignite::{{closure}}::hee55e51c137ee882
                               at /tmp/mobilenode/.mob/cargo/registry/src/github.com-1ecc6299db9ec823/rocket-0.5.0-rc.2/src/rocket.rs:540:9
                           <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll::h83de656ef4091c14
                               at /rustc/e85edd9a844b523a02dbd89f3c02cd13e4c9fe46/library/core/src/future/mod.rs:91:19
                           rocket::rocket::Rocket<P>::launch::{{closure}}::heb042593b03646e4
                               at /tmp/mobilenode/.mob/cargo/registry/src/github.com-1ecc6299db9ec823/rocket-0.5.0-rc.2/src/rocket.rs:893:56
                           <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll::h4a73f71be8381599
                               at /rustc/e85edd9a844b523a02dbd89f3c02cd13e4c9fe46/library/core/src/future/mod.rs:91:19
                           wallet_service_mirror_public::main::{{closure}}::ha4a350ed35bc0ec5
                               at /tmp/mobilenode/mirror/src/public/main.rs:231:16
                           <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll::h5afae002a1529c07
                               at /rustc/e85edd9a844b523a02dbd89f3c02cd13e4c9fe46/library/core/src/future/mod.rs:91:19
  13:     0x558c7d18b48d - tokio::park::thread::CachedParkThread::block_on::{{closure}}::h0c3081521f871928
                               at /tmp/mobilenode/.mob/cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.19.2/src/park/thread.rs:263:54
                           tokio::coop::with_budget::{{closure}}::h6051f536cb820bf6
                               at /tmp/mobilenode/.mob/cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.19.2/src/coop.rs:102:9
                           std::thread::local::LocalKey<T>::try_with::ha9625cc84902a721
                               at /rustc/e85edd9a844b523a02dbd89f3c02cd13e4c9fe46/library/std/src/thread/local.rs:443:16
                           std::thread::local::LocalKey<T>::with::h05deee08c3ac5acd
                               at /rustc/e85edd9a844b523a02dbd89f3c02cd13e4c9fe46/library/std/src/thread/local.rs:419:9
  14:     0x558c7d1b743f - tokio::coop::with_budget::hafd9e71c16af8362
                               at /tmp/mobilenode/.mob/cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.19.2/src/coop.rs:95:5
                           tokio::coop::budget::hd07ded997fadb241
                               at /tmp/mobilenode/.mob/cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.19.2/src/coop.rs:72:5
                           tokio::park::thread::CachedParkThread::block_on::hdf6d532d4800411b
                               at /tmp/mobilenode/.mob/cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.19.2/src/park/thread.rs:263:31
  15:     0x558c7d1178dd - tokio::runtime::enter::Enter::block_on::h6723365d3e62dea0
                               at /tmp/mobilenode/.mob/cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.19.2/src/runtime/enter.rs:151:13
                           tokio::runtime::thread_pool::ThreadPool::block_on::h372b222e91285f89
                               at /tmp/mobilenode/.mob/cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.19.2/src/runtime/thread_pool/mod.rs:90:9
  16:     0x558c7d12c0ae - tokio::runtime::Runtime::block_on::ha8ff20bda2ea2516
                               at /tmp/mobilenode/.mob/cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.19.2/src/runtime/mod.rs:482:43
  17:     0x558c7d11823e - rocket::async_run::h72fc269648316da4
                               at /tmp/mobilenode/.mob/cargo/registry/src/github.com-1ecc6299db9ec823/rocket-0.5.0-rc.2/src/lib.rs:246:18
                           rocket::async_main::hf21fc090a21496ca
                               at /tmp/mobilenode/.mob/cargo/registry/src/github.com-1ecc6299db9ec823/rocket-0.5.0-rc.2/src/lib.rs:267:5
  18:     0x558c7d1c2b55 - wallet_service_mirror_public::main::h99466f52fb26e21d
                               at /tmp/mobilenode/mirror/src/public/main.rs:231:30
  19:     0x558c7d1a3063 - core::ops::function::FnOnce::call_once::hbfe6a9ea98588ac6
                               at /rustc/e85edd9a844b523a02dbd89f3c02cd13e4c9fe46/library/core/src/ops/function.rs:248:5
                           std::sys_common::backtrace::__rust_begin_short_backtrace::h42313ef7581cae40
                               at /rustc/e85edd9a844b523a02dbd89f3c02cd13e4c9fe46/library/std/src/sys_common/backtrace.rs:122:18
  20:     0x558c7d208a29 - std::rt::lang_start::{{closure}}::ha31502fd1e5f710c
                               at /rustc/e85edd9a844b523a02dbd89f3c02cd13e4c9fe46/library/std/src/rt.rs:145:18
  21:     0x558c7dbd07ee - core::ops::function::impls::<impl core::ops::function::FnOnce<A> for &F>::call_once::hb78536f761ac4ccf
                               at /rustc/e85edd9a844b523a02dbd89f3c02cd13e4c9fe46/library/core/src/ops/function.rs:280:13
                           std::panicking::try::do_call::h0a4e2429dcf6aa05
                               at /rustc/e85edd9a844b523a02dbd89f3c02cd13e4c9fe46/library/std/src/panicking.rs:492:40
                           std::panicking::try::ha28d384da4d3830b
                               at /rustc/e85edd9a844b523a02dbd89f3c02cd13e4c9fe46/library/std/src/panicking.rs:456:19
                           std::panic::catch_unwind::h5128220fb91881a2
                               at /rustc/e85edd9a844b523a02dbd89f3c02cd13e4c9fe46/library/std/src/panic.rs:137:14
                           std::rt::lang_start_internal::{{closure}}::h93d0c63c493948b2
                               at /rustc/e85edd9a844b523a02dbd89f3c02cd13e4c9fe46/library/std/src/rt.rs:128:48
                           std::panicking::try::do_call::hed0daa62200c4fcd
                               at /rustc/e85edd9a844b523a02dbd89f3c02cd13e4c9fe46/library/std/src/panicking.rs:492:40
                           std::panicking::try::h19cdb5136875f739
                               at /rustc/e85edd9a844b523a02dbd89f3c02cd13e4c9fe46/library/std/src/panicking.rs:456:19
                           std::panic::catch_unwind::h1df1a585997fa257
                               at /rustc/e85edd9a844b523a02dbd89f3c02cd13e4c9fe46/library/std/src/panic.rs:137:14
                           std::rt::lang_start_internal::h191f82060be993ea
                               at /rustc/e85edd9a844b523a02dbd89f3c02cd13e4c9fe46/library/std/src/rt.rs:128:20
  22:     0x558c7d1c2b88 - main
  23:     0x7f5cbf629d90 - <unknown>
  24:     0x7f5cbf629e40 - __libc_start_main
  25:     0x558c7d0c989e - _start
  26:                0x0 - <unknown>
```

Reply:
> It's only at the end when you shut down the app that this happens

How are you shutting it down?

Reply:
ctrl-c